254, 246: Update adapters documentation (#253)
insight-platform/Savant: docs/source/savant_101/10_adapters.rst

    sub+bind:tcp://0.0.0.0:3331
    pub+connect:tcp://10.0.0.10:3331
    # etc

When the transports are specified  with the environment variables it looks like:

.. code-block:: bash

    # Unix domain socket communication
    ZMQ_ENDPOINT="dealer+connect:ipc:///tmp/zmq-sockets/input-video.ipc"

    # which is equal to
    ZMQ_ENDPOINT="ipc:///tmp/zmq-sockets/input-video.ipc"
    ZMQ_TYPE="DEALER"
    ZMQ_BIND="False"

Or:

.. code-block:: bash

    # tcp socket communication
    ZMQ_ENDPOINT="pub+bind:tcp://1.1.1.1:3333"

    # which is equal to
    ZMQ_ENDPOINT="tcp://1.1.1.1:3333"
    ZMQ_TYPE="PUB"
    ZMQ_BIND="True"

Not all socket pairs form "sane" communication patterns, so, you must use combinations colored green:

.. image:: ../_static/img/10_adapters_normal_pairs.png

The Rules Of Thumb
------------------

Consider the following ideas when planning ``source-framework-sink`` topologies:

- Use the framework in the bind mode, adapters in the connect mode; change if it does not work for you.
- The party which delivers multiplexed stream usually has the bind type; the party which handles a single (non-multiplexed) stream usually has the connect type.

> - Use the ``PUB/SUB`` pair when the pipeline or adapter is capable to handle the traffic in real-time.

Reviewer: ```suggestion
- Use the ``PUB/SUB`` pair when the pipeline or adapter is capable of handling the traffic in real-time.
```

Reviewer: or `able to handle`

The second is typical when adapters deliver data from several sources (e.g. RTSP cams) into a framework instance. The right side of the pipeline stays the same as in the previous case.

Edge Patterns
^^^^^^^^^^^^^

Edge is usually used to serve low-latency real-time video processing. To implement that, we establish the ``PUB/SUB`` connection because it drops the packets that the ``SUB`` part cannot process on time.

This mechanism works great with streams delivering ``MJPEG``, ``RAW``, ``JPEG``, ``PNG``, and other independently encoded video frames. Using it with keyframe-encoded streams leads to video corruption.

.. image:: ../_static/img/10_adapters_edge_patterns.png

The first pattern can be used when neither adapters nor the framework must get stuck because of the sink stalling. The second pattern is beneficial when a sink guarantees processing, and you do not worry that it may cause stalling.

DEALER/ROUTER
^^^^^^^^^^^^^

> This is a recommended pair when you don't need to copy the same messages to multiple subscribers. It is a reliable socket pair: the ``DEALER`` will block if the ``ROUTER``'s queue is full.

Reviewer: ```suggestion
This is the recommended pair when you don't need to copy the same messages to multiple subscribers. It is a reliable socket pair: the ``DEALER`` will block if the ``ROUTER``'s queue is full.
```

**Source/CONNECT, Framework/BIND**. This is a typical scheme.

.. image:: ../_static/img/10_adapters_dr_scfb.png

**Framework/CONNECT, Sink/BIND**. This is a normal pattern when a sink adapter communicates with an external system like Kafka and wishes to send data from multiple framework instances.

.. image:: ../_static/img/10_adapters_dr_fcsb.png

> **Source/BIND, Framework/CONNECT**. This is an exotic pattern. Nevertheless, it does the job when a module handles independent images without the need to maintain per-source order. In this scheme, the source will evenly distribute data between connected frameworks according to the ``LRU`` strategy, so it is impossible to use the scheme when you work with video.

Reviewer: This section seems to use the terms `framework` / `module` interchangeably.

I think it's better to settle on a single term. In my opinion, `module` is the better one, since the page begins with its definition.

Author: Fixed

.. image:: ../_static/img/10_adapters_dr_sbfc.png

**Framework/BIND, Sink/CONNECT**. This is a valid pattern when sinks communicating with an external system require partitioning and data appending order is not critical.

.. image:: ../_static/img/10_adapters_dr_fbsc.png

REQ/REP
^^^^^^^

The ``REQ/REP`` pair is similar to ``DEALER/ROUTER`` except that the ``REQ`` part receives replies from the ``REP`` part every time the ``REP`` part reads the message.

PUB/SUB
^^^^^^^

The ``PUB/SUB`` is convenient when you need to duplicate the same data to multiple subscribers. Another use case is real-time data processing: excessive elements are dropped if the pipeline cannot handle the traffic.

> **Source/BIND, Framework/CONNECT**. A source is initialized as a server (bind), and a framework connects to it. This scheme can be used when the source already delivers multiple streams or the framework handles a single stream provided by the source. In this scenario, the source can duplicate the same stream to multiple frameworks simultaneously.

Reviewer: If the suggestion above is accepted, then `a framework` -> `a module`.

`a framework` -> `a framework instance` otherwise.

Author: Fixed

.. image:: ../_static/img/10_adapters_ps_sbfc.png

> **Framework/BIND, Sink/CONNECT**. This scheme is used widely. A framework duplicates the same data to multiple sinks. A sink can filter out only required data.

Reviewer: Suggest `A framework` -> `a module` / `a framework instance`

Author: Fixed

.. image:: ../_static/img/10_adapters_ps_fbsc.png

> **Source/CONNECT, Framework/BIND**. A typical scheme when a framework handles multiple streams. The framework binds to a socket and adapters connect to that socket.

Reviewer: Suggest `A framework` -> `a module` / `a framework instance`

Author: Fixed

.. image:: ../_static/img/10_adapters_ps_scfb.png

> **Framework/CONNECT, Sink/BIND**. This is unusual but a correct scheme. A sink handles multiple outputs from frameworks to deliver them in a storage, e.g. Kafka or ClickHouse.

Reviewer: ```suggestion
**Framework/CONNECT, Sink/BIND**. This is an unusual but correct scheme. A sink handles multiple outputs from modules to deliver them into a storage, e.g. Kafka or ClickHouse.
```

Author: Fixed
